## Investigating Airplane Accidents & Analyzing Time Complexity

We're going to investigate airplane accidents using a data.gov [dataset](https://catalog.data.gov/dataset/aviation-data-and-documentation-from-the-ntsb-accident-database-system-05748) from 2015. To do so I will:

* Read the data from a .txt file into a variety of arrays
* See how long it takes to find particular incidents
* Perform a variety of recursion searches to see what is most efficient

In [0]:
aviation_data = []
aviation_list = []
with open("AviationData.txt") as file:
  for line in file:
    line = line.strip()
    aviation_data.append(line)
    text = line.split(" | ")
    aviation_list.append(text)



In [143]:
aviation_data[0:2]

['Event Id | Investigation Type | Accident Number | Event Date | Location | Country | Latitude | Longitude | Airport Code | Airport Name | Injury Severity | Aircraft Damage | Aircraft Category | Registration Number | Make | Model | Amateur Built | Number of Engines | Engine Type | FAR Description | Schedule | Purpose of Flight | Air Carrier | Total Fatal Injuries | Total Serious Injuries | Total Minor Injuries | Total Uninjured | Weather Condition | Broad Phase of Flight | Report Status | Publication Date |',
 '20150908X74637 | Accident | CEN15LA402 | 09/08/2015 | Freeport, IL | United States | 42.246111 | -89.581945 | KFEP | albertus Airport | Non-Fatal | Substantial | Unknown | N24TL | CLARKE REGINALD W | DRAGONFLY MK |  |  |  | Part 91: General Aviation |  | Personal |  |  | 1 |  |  | VMC | TAKEOFF | Preliminary | 09/09/2015 |']

In [144]:
aviation_list[0:2]

[['Event Id',
  'Investigation Type',
  'Accident Number',
  'Event Date',
  'Location',
  'Country',
  'Latitude',
  'Longitude',
  'Airport Code',
  'Airport Name',
  'Injury Severity',
  'Aircraft Damage',
  'Aircraft Category',
  'Registration Number',
  'Make',
  'Model',
  'Amateur Built',
  'Number of Engines',
  'Engine Type',
  'FAR Description',
  'Schedule',
  'Purpose of Flight',
  'Air Carrier',
  'Total Fatal Injuries',
  'Total Serious Injuries',
  'Total Minor Injuries',
  'Total Uninjured',
  'Weather Condition',
  'Broad Phase of Flight',
  'Report Status',
  'Publication Date |'],
 ['20150908X74637',
  'Accident',
  'CEN15LA402',
  '09/08/2015',
  'Freeport, IL',
  'United States',
  '42.246111',
  '-89.581945',
  'KFEP',
  'albertus Airport',
  'Non-Fatal',
  'Substantial',
  'Unknown',
  'N24TL',
  'CLARKE REGINALD W',
  'DRAGONFLY MK',
  '',
  '',
  '',
  'Part 91: General Aviation',
  '',
  'Personal',
  '',
  '',
  '1',
  '',
  '',
  'VMC',
  'TAKEOFF',
  'Preli

## For Loop

Below we'll do a nested for loop to find the item we care about, "LAX94LA336"


In [145]:
%%timeit
lax_code = []
for row in aviation_list:
  for item in row:
    if "LAX94LA336" in item:
      lax_code.append(row)


10 loops, best of 3: 92 ms per loop


In [146]:
lax_code

[['20001218X45447 ',
  ' Accident ',
  ' LAX94LA336 ',
  ' 07/19/1962 ',
  ' BRIDGEPORT, CA ',
  ' United States ',
  '  ',
  '  ',
  '  ',
  '  ',
  ' Fatal(4) ',
  ' Destroyed ',
  '  ',
  ' N5069P ',
  ' PIPER ',
  ' PA24-180 ',
  ' No ',
  ' 1 ',
  ' Reciprocating ',
  '  ',
  '  ',
  ' Personal ',
  '  ',
  ' 4 ',
  ' 0 ',
  ' 0 ',
  ' 0 ',
  ' UNK ',
  ' UNKNOWN ',
  ' Probable Cause ',
  ' 09/19/1996 ',
  ' \n']]

## Results from the For Loop

While the for loop above does the job, it takes multiple nested for loops to get through every row and then every item in the row to find the string we care about. 

This results in a time complexity of O(m*n) or quadratic time, with m being the number of rows and n being the number of items in each row, O(n^2).

This results in a time of **10 loops, with each loop taking 96 ms**.


## Linear Algorithm

Below we'll do write an algorithm that uses linear time to find the value we care about, "LAX94LA336".

This will result in time complexity of O(n) as the dataset gets larger, time increases linearly


In [0]:
lax_code_2 = []
def find_accident_number(data, accident_id):
  for row in data:
    if row[2] == accident_id:
      lax_code_2.append(row)
    else:
      pass

In [148]:
%%timeit
find_accident_number(aviation_list, 'LAX94LA336')

10 loops, best of 3: 26.9 ms per loop


In [149]:
lax_code_2

[['20001218X45447',
  'Accident',
  'LAX94LA336',
  '07/19/1962',
  'BRIDGEPORT, CA',
  'United States',
  '',
  '',
  '',
  '',
  'Fatal(4)',
  'Destroyed',
  '',
  'N5069P',
  'PIPER',
  'PA24-180',
  'No',
  '1',
  'Reciprocating',
  '',
  '',
  'Personal',
  '',
  '4',
  '0',
  '0',
  '0',
  'UNK',
  'UNKNOWN',
  'Probable Cause',
  '09/19/1996 |'],
 ['20001218X45447',
  'Accident',
  'LAX94LA336',
  '07/19/1962',
  'BRIDGEPORT, CA',
  'United States',
  '',
  '',
  '',
  '',
  'Fatal(4)',
  'Destroyed',
  '',
  'N5069P',
  'PIPER',
  'PA24-180',
  'No',
  '1',
  'Reciprocating',
  '',
  '',
  'Personal',
  '',
  '4',
  '0',
  '0',
  '0',
  'UNK',
  'UNKNOWN',
  'Probable Cause',
  '09/19/1996 |'],
 ['20001218X45447',
  'Accident',
  'LAX94LA336',
  '07/19/1962',
  'BRIDGEPORT, CA',
  'United States',
  '',
  '',
  '',
  '',
  'Fatal(4)',
  'Destroyed',
  '',
  'N5069P',
  'PIPER',
  'PA24-180',
  'No',
  '1',
  'Reciprocating',
  '',
  '',
  'Personal',
  '',
  '4',
  '0',
  '0',


## Results from Linear Function

Doing one for loop increased reduced our overall time per loop to **18.3 ms per loop**.

This results in a time complexity of O(n) or linear time, with n being the number of loops before finding the value we care about.

## Logarithmic Algorithm

Below we'll do write an algorithm that uses logarithmic time to find the value we care about, "LAX94LA336".

This will result in time complexity of O(log n) as the dataset gets larger, time increases logarithmically.

However we'll have to have our list be ordered by accident ID first.


In [0]:
sorted_aviation_list = sorted(aviation_list, key= lambda row: row[2])




In [151]:
sorted_aviation_list[0]

['20001212X20184',
 'Accident',
 'ANC00FA018',
 '12/07/1999',
 'BETHEL, AK',
 'United States',
 '',
 '',
 '',
 '',
 'Fatal(6)',
 'Destroyed',
 '',
 'N1747U',
 'Cessna',
 '207',
 'No',
 '1',
 'Reciprocating',
 '',
 'SCHD',
 'Unknown',
 'GRANT AVIATION, INC.',
 '6',
 '0',
 '0',
 '0',
 'IMC',
 'CRUISE',
 'Probable Cause',
 '04/18/2001 |']

In [0]:
import math
def find_accident_information(data, accident_id):
    length = len(data)
    upper_bound = length - 1
    lower_bound = 0

    index = math.floor((upper_bound + lower_bound) / 2)

    guess = data[index][2]
    accident_info = data[index]

    while accident_id != guess:
        if accident_id < guess:
          upper_bound = index - 1
        else:
          lower_bound = index + 1
        index = math.floor((upper_bound + lower_bound) / 2)
        guess = data[index][2]
        accident_info = data[index]

    return accident_info


In [153]:
%%timeit
find_accident_information(sorted_aviation_list, "LAX94LA336")

The slowest run took 5.55 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 4.8 µs per loop


## Results from Logarithmic Function

Doing Binary Search on the accident id resulted in a much reduced time of 4.75 microseconds per loop

This results in a time complexity of O(log n) or logarithmic time

## Creating a dictionary

Below we'll create a list of data headers, and go through each row and create a dictionary using the dict(zip()) method on each split row

In [154]:
# While we can get the list of headers, the last one doesn't quite format correctly
aviation_dict_list = []
header_list = []
header = aviation_data[0].split(" | ")

header

['Event Id',
 'Investigation Type',
 'Accident Number',
 'Event Date',
 'Location',
 'Country',
 'Latitude',
 'Longitude',
 'Airport Code',
 'Airport Name',
 'Injury Severity',
 'Aircraft Damage',
 'Aircraft Category',
 'Registration Number',
 'Make',
 'Model',
 'Amateur Built',
 'Number of Engines',
 'Engine Type',
 'FAR Description',
 'Schedule',
 'Purpose of Flight',
 'Air Carrier',
 'Total Fatal Injuries',
 'Total Serious Injuries',
 'Total Minor Injuries',
 'Total Uninjured',
 'Weather Condition',
 'Broad Phase of Flight',
 'Report Status',
 'Publication Date |']

In [155]:
for item in header:
  item = item.strip("|").strip()
  header_list.append(item)

header_list

['Event Id',
 'Investigation Type',
 'Accident Number',
 'Event Date',
 'Location',
 'Country',
 'Latitude',
 'Longitude',
 'Airport Code',
 'Airport Name',
 'Injury Severity',
 'Aircraft Damage',
 'Aircraft Category',
 'Registration Number',
 'Make',
 'Model',
 'Amateur Built',
 'Number of Engines',
 'Engine Type',
 'FAR Description',
 'Schedule',
 'Purpose of Flight',
 'Air Carrier',
 'Total Fatal Injuries',
 'Total Serious Injuries',
 'Total Minor Injuries',
 'Total Uninjured',
 'Weather Condition',
 'Broad Phase of Flight',
 'Report Status',
 'Publication Date']

In [0]:
for item in aviation_data[1:]:
    split_row = item.split(" | ")
    dict_row = dict(zip(header, split_row))
    aviation_dict_list.append(dict_row)

In [157]:
aviation_dict_list[0]

{'Accident Number': 'CEN15LA402',
 'Air Carrier': '',
 'Aircraft Category': 'Unknown',
 'Aircraft Damage': 'Substantial',
 'Airport Code': 'KFEP',
 'Airport Name': 'albertus Airport',
 'Amateur Built': '',
 'Broad Phase of Flight': 'TAKEOFF',
 'Country': 'United States',
 'Engine Type': '',
 'Event Date': '09/08/2015',
 'Event Id': '20150908X74637',
 'FAR Description': 'Part 91: General Aviation',
 'Injury Severity': 'Non-Fatal',
 'Investigation Type': 'Accident',
 'Latitude': '42.246111',
 'Location': 'Freeport, IL',
 'Longitude': '-89.581945',
 'Make': 'CLARKE REGINALD W',
 'Model': 'DRAGONFLY MK',
 'Number of Engines': '',
 'Publication Date |': '09/09/2015 |',
 'Purpose of Flight': 'Personal',
 'Registration Number': 'N24TL',
 'Report Status': 'Preliminary',
 'Schedule': '',
 'Total Fatal Injuries': '',
 'Total Minor Injuries': '',
 'Total Serious Injuries': '1',
 'Total Uninjured': '',
 'Weather Condition': 'VMC'}

In [158]:
%%timeit
lax_dict = []
for row in aviation_dict_list:
    if "LAX94LA336" in row.values():
        lax_dict.append(row)

10 loops, best of 3: 66.8 ms per loop


In [159]:
lax_dict

[{'Accident Number': 'LAX94LA336',
  'Air Carrier': '',
  'Aircraft Category': '',
  'Aircraft Damage': 'Destroyed',
  'Airport Code': '',
  'Airport Name': '',
  'Amateur Built': 'No',
  'Broad Phase of Flight': 'UNKNOWN',
  'Country': 'United States',
  'Engine Type': 'Reciprocating',
  'Event Date': '07/19/1962',
  'Event Id': '20001218X45447',
  'FAR Description': '',
  'Injury Severity': 'Fatal(4)',
  'Investigation Type': 'Accident',
  'Latitude': '',
  'Location': 'BRIDGEPORT, CA',
  'Longitude': '',
  'Make': 'PIPER',
  'Model': 'PA24-180',
  'Number of Engines': '1',
  'Publication Date |': '09/19/1996 |',
  'Purpose of Flight': 'Personal',
  'Registration Number': 'N5069P',
  'Report Status': 'Probable Cause',
  'Schedule': '',
  'Total Fatal Injuries': '4',
  'Total Minor Injuries': '0',
  'Total Serious Injuries': '0',
  'Total Uninjured': '0',
  'Weather Condition': 'UNK'}]

## Overall thoughts on list of lists versus list of dict

Overall I'd say the complexity of using a list of list versus list of dictionary is relatively higher. The dictionary has the advantage of being able to search and organization keys (not just indices)

## Counting number of accidents per state

We're going to count the number of accidents by state, to do so we'll have to:

1. Parse the state from the "Location" column
1. Iterate and/or keep track of the number of accidents per state in some fashion

In [0]:
state_list = []

for int in range(0, len(aviation_dict_list)):
  if aviation_dict_list[int]["Country"] == "United States":
    location = aviation_dict_list[int]["Location"]
    state = location.split(" ")[-1]
    if state != "":
      state_list.append(state)

In [0]:
from collections import Counter

state_accidents = Counter(state_list)

In [176]:
state_accidents.most_common(1)

[('CA', 8030)]

## Results

California had the largest number of accidents, with 8030 based on our counter of all the incidents in the US where states were accounted for.


## Determine The Month & Year Fatalities and Injuries

We will count how many fatalities and serious injuries occured during each unique month and year and what the largest # of fatalities and injury was by month and year.


In [0]:
def calculate_monthly_pain(data):
    monthly_injuries = {}
    for x in range(0, len(aviation_dict_list)):
        total_injuries = 0
        year = aviation_dict_list[x]["Event Date"].split("/")[-1]
        month = aviation_dict_list[x]["Event Date"].split("/")[0]
        deaths = aviation_dict_list[x]["Total Fatal Injuries"]
        serious_injuries = aviation_dict_list[x]["Total Serious Injuries"] 

        if len(year) == 4 and len(month) == 2:
            year_month = year + "-" + month
        if deaths == '':
            deaths = 0
        if serious_injuries == '':
            serious_injuries = 0 

        if year_month not in monthly_injuries.keys():
            monthly_injuries[year_month] = (float(deaths) + float(serious_injuries))
        elif year_month in monthly_injuries.keys():
            monthly_injuries[year_month] =+ (float(deaths) + float(serious_injuries))

    return monthly_injuries

In [0]:
from collections import Counter

monthly_injuries = calculate_monthly_pain(aviation_dict_list)

monthly_hurt = Counter(monthly_injuries)

In [286]:
monthly_hurt.most_common(1)

[('2007-01', 102.0)]